File: STUDY/2/3.pas
var
 n,m,max,min,l,r,mid,tmp,ans:int64;
 i:longint;
 a:array[1..100000] of longint;
begin
	read(n,m);
	min:=100000;
	for i:=1 to n do begin read(a[i]); if a[i]>max then max:=a[i]; if a[i]<min then min:=a[i]; end;
	l:=min;
	r:=max*m;
	while (l<=r) do
	  begin
	  	mid:=(l+r) div 2;
	  	tmp:=0;
	  	for i:=1 to n do tmp:=tmp+(mid div a[i]);
	  	if tmp>=m then begin r:=mid-1; ans:=mid; end
	  	 else l:=mid+1;
	  end;
        write(ans);
end.

File: STUDY/21/3.cpp
#include <bits/stdc++.h>
using namespace std;

int main(){
	int m,n,a[100005],tmp,maxl=0,ans=0;
	cin>>m>>n;
	for (int i=1;i<=m;++i){
		scanf("%d",&a[i]);
		maxl=max(maxl,a[i]);
	}
	int l=1,r=maxl,mid;
	while (l<=r){
		mid=(l+r)/2;
		tmp=0;
		for (int i=1;i<=m;++i)
			tmp+=a[i]/mid;
		if (tmp>=n){
			l=mid+1;
			ans=mid;
		}else{
			r=mid-1;
		}
	}
	cout<<ans<<endl;
	return 0;
}